www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,126 @@
+5. **Ask the transaction coordinator to distribute the locks**
+
+The reason we extract the locks and distribute them early in the process is to minimize the time spent holding a lock. Most DDL statements require an `ACCESS EXCLUSIVE` lock, which not only prevents any queries from accessing the table in question while the DDL is executing but also prevents any new queries from accessing the table while the DDL is waiting to acquire the lock. To minimize the risk and the time spent while holding the lock, Nile attempts to acquire the necessary locks - with a short lock timeout - on all relevant databases before starting to execute the DDLs. If the lock acquisition fails on any database, the DDL will return an error rather than continue waiting for the lock.
+
+Acquiring locks with a short timeout before executing DDLs is considered a best practice in Postgres for the reasons we just explained. Nile's distributed DDL implement this best practice for our users.
+
+### Transaction coordinator
+
+Once the `nile_ddl` extension determines the necessary locks, it calls the transaction coordinator to start a transaction (if necessary) and distribute the locks.
+
+Starting the transaction is straightforward: the coordinator maintains open connections to all databases and simply sends each one a `BEGIN` command. 
+
+Distributing the locks works similarly. The coordinator sends all databases the commands required to acquire the locks (more details in the section on locking). To avoid deadlocks, we ensure that locks are always acquired in the exact same order - both in terms of the sequence of databases and the order of locks within each database. This guarantees early failure in the event of conflicts and prevents situations where two concurrent transactions are each waiting for the other to release a lock on a different database.
+
+![Distributing locks](/blog/2024-11-19-distributed-ddl/distribute_locks.png)
+
+Once the locks are acquired, the `nile_ddl` extension instructs the coordinator to distribute the DDL itself. While the coordinator sends the DDL command to all other databases, the original extension that first recieved the DDL proceeds to call `standard_ProcessUtility` and process the DDL locally.
+
+Meanwhile, each remote database recieves the DDL command. Since every database runs our extension, these DDL commands are intercepted by the extension on each database. It is crucial that the extension does not attempt to redistribute these DDLs, as doing so would lead to an infinite loop. Therefore, the extension has to recognize that these DDLs were sent by the transaction coordinator and can be passed directly to `standard_ProcessUtility`. To achieve this, we use a configuration (GUC) set when the transaction coordinator initializes the transaction. This configuration indicates to the extension that it doesn't need to reprocess the DDLs that follow, as they have already been validated by the originating extension and distributed by the coordinator.
```

Comment:
another problem that can happen is that the orignal DDL could be ambiguous.

for instance, if you create an unqualified table, there's no guarantee that the statement would lead to the exact same object on all nodes depending on concurrent DDL with something as simple as sending the search_path. if it contains a trailing unexisting schema that is concurrently created, maybe one of the execution could happen after the new schema is created while other happen before. in such a scenario you would end up with a successful creation of the wrong object.

so an important role of the extension is to first make the DDL entirely unambiguous so we can guarantee the exact same behaviour on all the nodes.

Reply:
Thanks. I had it in my talk notes, but I figured I can leave it out for simplicity. You are right that its important and also doesn't really complicate the description.